lib/cstring.cpp

```diff
@@ -166,9 +166,17 @@ struct TableEntryHash {
 auto &cache() {
     // We need node_hash_set due to SSO: we return address of embedded string
     // that should be stable
-    static absl::node_hash_set<table_entry, TableEntryHash, std::equal_to<>> g_cache;
 
-    return g_cache;
+    // Note that we need to lazily initialize on the first use: static
+    // initialization order is undefined and we need cache to be live *before*
+    // and static cstrings around.
+    // FIXME: Do something similar to LLVM's ManagedStatic to be able to destruct
+    // this object on exit.
+    using CacheType = absl::node_hash_set<table_entry, TableEntryHash, std::equal_to<>>;
+    static CacheType *g_cache = nullptr;
+    if (!g_cache) g_cache = new CacheType;
+
+    return *g_cache;
```

Comment:
static locals are defined to be initialized the first time the containing function is called, so there's no need to do this.  Indeed, it is not (generally) safe to call new/malloc from a function that might be called from a static global ctor, so this is not safe.